### PR TITLE
Desktop package and test app

### PR DESCRIPTION
**Pull request by kckst8**
The desktop-viewer-test app will be re-purposed as the new CRA template for desktop (v1) once this is approved/merged. The next iteration will add true offline briefcase support with download/sync.

Will be adding unit tests for the package as this is being reviewed but wanted to get eyes on this ASAP in hopes of getting the CRA template out in Q2

### CODE REVIEW

File: packages/apps/cdn-viewer/package.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
     "@bentley/presentation-frontend": "^2.14.0",
     "@bentley/product-settings-client": "^2.14.0",
     "@bentley/rbac-client": "^2.14.0",
-    "@bentley/react-scripts": "3.4.1",
+    "@bentley/react-scripts": "3.4.9",
```
Comment (aruniverse) [inline packages/apps/cdn-viewer/package.json:24]: i would actually bump this to 3.4.11 throughout the monorepo

File: .vscode/launch.json [hunk 2]
```diff
@@ -40,5 +50,34 @@
...
+      "name": "[BACKEND] itwin-viewer-desktop",
+      "cwd": "${workspaceFolder}/packages/apps/desktop-viewer-test",
+      "type": "pwa-node",
+      "request": "launch",
+      "console": "internalConsole",
+      "outputCapture": "std",
+      "runtimeExecutable": "${workspaceFolder}/common/temp/node_modules/.bin/electron",
+      "runtimeArgs": [
+        "${workspaceFolder}/packages/apps/desktop-viewer-test/lib/backend/main.js",
+        "--remote-debugging-port=9223"
```
Comment (aruniverse) [inline .vscode/launch.json:63]: itll have the same debugging port as the frontend?

File: packages/modules/error-handling-react/package.json [hunk 3]
```diff
@@ -56,7 +56,6 @@
     "@bentley/ui-core": "^2.14.0",
     "@svgr/cli": "^5.3.1",
     "@testing-library/react": "^10.4.4",
-    "@types/classnames": "^2.2.10",
```
Comment (aruniverse) [inline packages/modules/error-handling-react/package.json:59]: i would bump classnames to v^2.3.1 if youre removing the typedefs, that ver should have the types for the pkg

File: packages/modules/viewer-react/src/components/BaseViewer.tsx [hunk 4]
```diff
@@ -41,24 +41,30 @@ export const BaseViewer: React.FC<ViewerProps> = ({
   additionalI18nNamespaces,
   additionalRpcInterfaces,
 }: ViewerProps) => {
-  const [authorized, setAuthorized] = useState(false);
+  // assume authorized when using a local snapshot
+  const [authorized, setAuthorized] = useState(snapshotPath ? true : false);
```
Comment (aruniverse) [inline packages/modules/viewer-react/src/components/BaseViewer.tsx:45]:
```suggestion
  const [authorized, setAuthorized] = useState(!!snapshotPath);
```

File: packages/apps/desktop-viewer-test/.env [hunk 5]
```diff
@@ -0,0 +1,4 @@
+# ---- Configuration ----
+ITWIN_VIEWER_SCOPE ="email rbac-user:external-client projectwise-share urlps-third-party general-purpose-imodeljs-backend product-settings-service context-registry-service:read-only imodelhub organization profile openid imodeljs-router"
```
Comment (aruniverse) [inline packages/apps/desktop-viewer-test/.env:2]: would the desktop even need to use gpb? doesnt ht ebackend run locally, so you wouldnt need that scope, `general-purpose-imodeljs-backend` ?
also, what about the prdss one?

File: packages/apps/desktop-viewer-test/src/backend/main.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (aruniverse) [inline packages/apps/desktop-viewer-test/src/backend/main.ts:1]: a lot of files have duplicate headers

File: packages/apps/desktop-viewer-test/src/frontend/components/widgets/PropertyGridWidget.tsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import { Orientation } from "@bentley/ui-core";
+import { ConfigurableCreateInfo, WidgetControl } from "@bentley/ui-framework";
+import * as React from "react";
+
+import SimplePropertiesComponent from "../Properties";
+
+/** A widget control for displaying the PropertyGrid React component */
+export class PropertyGridWidget extends WidgetControl {
```
Comment (aruniverse) [inline packages/apps/desktop-viewer-test/src/frontend/components/widgets/PropertyGridWidget.tsx:16]: why is this needed? doesnt the base-viewer already have the property-grid?

File: packages/apps/desktop-viewer-test/src/frontend/index.tsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (aruniverse) [inline packages/apps/desktop-viewer-test/src/frontend/index.tsx:1]: dup headers

File: packages/modules/desktop-viewer-react/src/services/Initializer.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ClientRequestContext } from "@bentley/bentleyjs-core";
+import {
+  ElectronApp,
+  ElectronAppOpts,
+} from "@bentley/electron-manager/lib/ElectronFrontend";
```
Comment (aruniverse) [inline packages/modules/desktop-viewer-react/src/services/Initializer.ts:10]: are these  really not exported in the main barrel?
i personally am not a big fan of imports from lib like this

File: packages/modules/desktop-viewer-react/src/services/Initializer.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    }
+    return {
+      info: {
+        title: backendOptions.hostedBackend.title,
+        version: backendOptions.hostedBackend.version,
+      },
+      uriPrefix: orchestratorUrl,
+    };
+  } else {
+    return {
```
Comment (aruniverse) [inline packages/modules/desktop-viewer-react/src/services/Initializer.ts:47]: i did make another comment about thsis in another file, but shouldnt desktop starter have a local backend that is used? and no need for gpb?

**Review by aruniverse — COMMENTED** [review]
first pass, need to go over rest of the pr (mostly just the test desktop viewer) tomorrow

File: packages/modules/viewer-react/src/components/BaseViewer.tsx (hunk 4 above)
Comment (kckst8) [reply packages/modules/viewer-react/src/components/BaseViewer.tsx]: Not a fan of double negation (less readable) but if you're going to be picky about it :-)

File: packages/apps/desktop-viewer-test/.env (hunk 5 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/.env:2]: there is still a "connected" mode in desktop mode where this will be used, but v2 with proper local briefcase support will likely warrant changes here

File: packages/modules/desktop-viewer-react/src/services/Initializer.ts (hunk 9 above)
Comment (kckst8) [reply packages/modules/desktop-viewer-react/src/services/Initializer.ts:10]: this is all that we have for these files. No barrel

File: .vscode/launch.json (hunk 2 above)
Comment (kckst8) [reply .vscode/launch.json:63]: this is the remote-debugging-port for the browser, so it needs to match

File: packages/apps/cdn-viewer/package.json (hunk 1 above)
Comment (kckst8) [reply packages/apps/cdn-viewer/package.json]: done

File: packages/modules/error-handling-react/package.json (hunk 3 above)
Comment (kckst8) [reply packages/modules/error-handling-react/package.json:59]: good catch. done

File: packages/apps/desktop-viewer-test/src/backend/main.ts (hunk 6 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/backend/main.ts:1]: fixed

File: packages/apps/desktop-viewer-test/src/frontend/components/widgets/PropertyGridWidget.tsx (hunk 7 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/frontend/components/widgets/PropertyGridWidget.tsx]: it's not. Copy leftover. I removed it

File: packages/apps/desktop-viewer-test/src/frontend/index.tsx (hunk 8 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/frontend/index.tsx:1]: fixed

File: packages/apps/desktop-viewer-test/src/frontend/app/ITwinViewerApp.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import {
```
Comment (kabentley) [inline packages/apps/desktop-viewer-test/src/frontend/app/ITwinViewerApp.ts:6]: please follow our `imports` style and not use one-per-line. Are you using our lint rules?

File: packages/apps/desktop-viewer-test/.env (hunk 5 above)
Comment (calebmshafer) [reply packages/apps/desktop-viewer-test/.env:2]: Not sure I understand that. The desktop mode should still not need GPB right?

File: packages/apps/desktop-viewer-test/package.json [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    "clean": "rimraf lib build dist",
+    "copy": "npm run copy:assets && npm run copy:enUS",
+    "copy:assets": "cpx \"./assets/*\" ./build/assets/",
+    "copy:enUS": "cpx \"./build/locales/en/*\" ./build/locales/en-US/",
+    "electron": "electron lib/backend/main.js",
+    "electron:debug": "cross-env NODE_ENV=development electron lib/backend/main.js",
+    "lint": "eslint --config package.json --no-eslintrc \"./src/**/*.{ts,tsx}\" 1>&2",
+    "start": "npm run build:backend && run-p \"start:frontend\" \"electron:debug\"",
+    "start:frontend": "cross-env USE_FULL_SOURCEMAPS=true USE_FAST_SASS=true BROWSER=none SKIP_PREFLIGHT_CHECK=true react-scripts start",
+    "build:pack": "electron-builder --dir",
```
Comment (calebmshafer) [inline packages/apps/desktop-viewer-test/package.json:26]: I'm hesitant to deliver this with electron-builder as our packager since we do have plans to switch it to something new.

Comment (aruniverse) [reply packages/apps/desktop-viewer-test/package.json:26]: this is the test app btw, so it should be ok as long as its not in the cra template

File: packages/apps/desktop-viewer-test/.env (hunk 5 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/.env:2]: currently it supports accessing models directly from the iTwin hub (similar to the desktop starter). Should this be removed?

File: packages/apps/desktop-viewer-test/package.json (hunk 12 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/package.json:26]: whatever is in here would be in the first version of the CRA template....this is just a starting point though. The plan is for v2 to be the more full featured desktop app (with local briefcase support). We can even make this a prerelease version if we think that makes sense.

@calebmshafer  how soon will we have a replacement? I'm hesitant to hold this up if we don't have a solution in place soon

File: packages/apps/desktop-viewer-test/README.md [hunk 13]
```diff
@@ -0,0 +1,28 @@
+# Desktop Viewer Sample
+
+This app is an example of using the iTwin Viewer in a modularized React + Electron desktop application. It is meant to be used as a development environment for the components in the packages in this monorepo. It is not intended to be used as a seed/starter application. You should use the [Create React App template](https://www.npmjs.com/package/@itwin/cra-template-desktop-viewer) to seed a new application.
+
+This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).
+
+## Configuration
```
Comment (aruniverse) [inline packages/apps/desktop-viewer-test/README.md:7]: So i hate to even suggest this, partially because its going to lead to more confusion, but maybe you want to add a link to the desktop starter tutorial or copy a lot of that in here.

Because when you register a new app you also need to setup the auth redirect uris. (I could be wrong about this, and this couldve changed with the auth changes keith made to be done in the backend) 
image

Comment (aruniverse) [reply packages/apps/desktop-viewer-test/README.md:7]: And if its not needed, we need to update developer.bentley to hide those fields

Comment (kckst8) [reply packages/apps/desktop-viewer-test/README.md:7]: The goals is for the CRA template to replace that tutorial (or be used in it) though. Once we have a CRA template, we will add more clearly defined steps to register your app (similar to the web CRA template README)

Comment (calebmshafer) [reply packages/apps/desktop-viewer-test/README.md:7]: The requirement for those redirects is still required so we should point to something.

File: packages/apps/desktop-viewer-test/src/frontend/app/ITwinViewerApp.ts (hunk 11 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/frontend/app/ITwinViewerApp.ts:6]: This repo uses prettier for formatting. Currently, it is using the default printWidth of 80 chars which will wrap long imports. I will update that to 120 in a separate PR since it will change pretty much every file and that should help with these (though it will not remove all of them). 

I've also setup a meeting to discuss unifying this and other open source repo linting/formatting rules and will reformat/refactor based on the outcome of that meeting

File: packages/apps/desktop-viewer-test/.env (hunk 5 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/.env:2]: my mistake....this code actually isn't being called. Leftover from web copy I suppose. I'll remove

Comment (kckst8) [reply packages/apps/desktop-viewer-test/.env:2]: removed

File: packages/apps/desktop-viewer-test/README.md (hunk 13 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/README.md:7]: this is just the developer app and it uses the default redirect uri:

`  public static defaultRedirectUri = "http://localhost:3000/signin-callback";
  private _configuration: AuthorizationServiceConfiguration | undefined;
  private _tokenResponse: TokenResponse | undefined;
  private _tokenStore?: ElectronTokenStore;
  public get tokenStore() { return this._tokenStore!; }

  public constructor(config?: NativeAppAuthorizationConfiguration) {
    super(config);
  }

  public get redirectUri() { return this.config?.redirectUri ?? ElectronAuthorizationBackend.defaultRedirectUri; }`

I don't think we need to add steps to change that since the intention of this app is strictly for developing features in the package. We can add additional steps to the CRA template.

File: packages/apps/desktop-viewer-test/package.json [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    "build:dist": "electron-builder",
+    "test": ""
```
Comment (aruniverse) [inline packages/apps/desktop-viewer-test/package.json:28]:
```suggestion
    "test": "",
    "pseudolocalize": "betools pseudolocalize --englishDir ./public/locales/en --out ./public/locales/en-PSEUDO"
```

File: packages/apps/desktop-viewer-test/src/frontend/components/AppComponent.tsx [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+    setConnectedMode(false);
+    setFrontstage("SnapshotSelector");
+    const frontstageDef =
+      FrontstageManager.findFrontstageDef("SnapshotSelector");
+    await FrontstageManager.setActiveFrontstageDef(frontstageDef);
+  };
+
+  const initIModel = useCallback(() => {
+    if (selectedIModel) {
+      setFrontstage("Viewer");
```
Comment (aruniverse) [inline packages/apps/desktop-viewer-test/src/frontend/components/AppComponent.tsx:104]: is it necessary to save these values to both state in the component and local storage?
ie instead of saved to app state, could just fetch the value from local storage with your fromLocalStorage fn

File: packages/apps/desktop-viewer-test/src/frontend/components/frontstages/SnapshotSelectFrontstage.tsx [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+      options
+    )) as OpenDialogReturnValue;
+    const file = val.canceled ? undefined : val.filePaths[0];
+    if (file) {
+      try {
+        ITwinViewerApp.store.dispatch({
+          type: "App:OPEN_SNAPSHOT",
+          payload: file,
+        });
+      } catch (e) {}
```
Comment (aruniverse) [inline packages/apps/desktop-viewer-test/src/frontend/components/frontstages/SnapshotSelectFrontstage.tsx:107]: maybe log the error?

File: packages/apps/desktop-viewer-test/src/frontend/components/frontstages/IModelSelectFrontstage.tsx [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+
+    this.reactNode = <IModelSelector onIModelSelected={this._onSelectIModel} />;
+  }
+
+  // called when an imodel has been selected on the IModelSelect
+  private _onSelectIModel = async (iModelInfo: IModelInfo) => {
+    ITwinViewerApp.store.dispatch({
+      type: "App:OPEN_IMODEL",
+      payload: {
+        projectName: iModelInfo.projectInfo.name,
```
Comment (aruniverse) [inline packages/apps/desktop-viewer-test/src/frontend/components/frontstages/IModelSelectFrontstage.tsx:39]: +1 to my other comment about if the names are even necessary

File: packages/apps/desktop-viewer-test/src/common/LoggerCategory.ts [hunk 18]
```diff
@@ -0,0 +1,10 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+/** List of LoggerCategories for this app.  For more details on Logging Categories, check out the [Category](https://www.imodeljs.org/learning/common/logging/#categories) documentation. */
+export enum AppLoggerCategory {
```
Comment (aruniverse) [inline packages/apps/desktop-viewer-test/src/common/LoggerCategory.ts:7]: How does logging work in a desktop app in production? Like are we writing out to a file or sending logs back to seq or something else via OpenTelemetry under the hood?

File: packages/apps/desktop-viewer-test/package.json [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+    "linux",
+    "darwin"
+  ],
+  "engines": {
+    "node": ">=10.17.0 <15.0"
+  },
+  "scripts": {
+    "build": "",
+    "build:backend": "tsc -p tsconfig.backend.json",
+    "build:frontend": "cross-env DISABLE_ESLINT=true SKIP_PREFLIGHT_CHECK=true GENERATE_SOURCEMAP=false USE_FAST_SASS=true TRANSPILE_DEPS=false DISABLE_TERSER=true react-scripts build",
```
Comment (aruniverse) [inline packages/apps/desktop-viewer-test/package.json:16]: i think itd be better if we moved these env vars from in the script into the .env file
something i wanted to change in core test-apps and desktop-starter too to be more consistent

File: packages/apps/desktop-viewer-test/public/locales/en-PSEUDO/iTwinViewer.json [hunk 20]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (aruniverse) [inline packages/apps/desktop-viewer-test/public/locales/en-PSEUDO/iTwinViewer.json:1]: could you run the new pseudolocalize comand and commit those changes?

File: packages/apps/desktop-viewer-test/src/backend/AppInfo.ts [hunk 21]
```diff
@@ -0,0 +1,12 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+export const appInfo = {
+  id: "itwin-viewer",
+  title: "iTwin Viewer",
```
[the remark this answers is not in the document]
Comment (aruniverse) [reply packages/apps/desktop-viewer-test/src/backend/AppInfo.ts:7]: wonder if we should be more explicit so people dont use them interchangeably when creating issues and discussion posts

File: packages/apps/desktop-viewer-test/src/common/ViewerConfig.ts [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+export interface ViewerConfig {
+  snapshotName?: string;
+  clientId: string;
+  redirectUri: string;
+}
+
+/** RPC interfaces required by the viewer */
+export const viewerRpcs = [
```
[the remark this answers is not in the document]
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/common/ViewerConfig.ts:29]: problem is currently the itv packages are designed to be frontend packages and we also are just using tsc to compile the backend (we would get errors when sass is encountered or any other file that can't be handled). We may end up needing "itwin-viewer-common" but I don't want to get into that yet for a change like this.

File: packages/apps/desktop-viewer-test/src/common/LoggerCategory.ts (hunk 18 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/common/LoggerCategory.ts:7]: in this app we are initializing to the console, but the logger class has different options that users could use to provide metadata, etc.

File: packages/apps/desktop-viewer-test/src/frontend/app/AppState.ts [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+// Valid values for SwitchIModelState.switchState
+export enum SwitchState {
+  None = 0,
+  SelectIModel = 1,
+  SelectSnapshot = 2,
+  OpenIModel = 3,
+  OpenSnapshot = 4,
+}
+
+export interface SelectedIModel {
```
[the remark this answers is not in the document]
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/frontend/app/AppState.ts:24]: nope. removed them

File: packages/apps/desktop-viewer-test/src/frontend/components/frontstages/SnapshotSelectFrontstage.tsx (hunk 16 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/frontend/components/frontstages/SnapshotSelectFrontstage.tsx]: done

File: packages/apps/desktop-viewer-test/package.json (hunk 19 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/package.json]: agree. done

File: packages/apps/desktop-viewer-test/public/locales/en-PSEUDO/iTwinViewer.json (hunk 20 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/public/locales/en-PSEUDO/iTwinViewer.json:1]: done

File: packages/apps/desktop-viewer-test/src/frontend/components/frontstages/IModelSelectFrontstage.tsx (hunk 17 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/frontend/components/frontstages/IModelSelectFrontstage.tsx]: removed

File: packages/apps/desktop-viewer-test/src/frontend/app/AppState.ts [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+ *--------------------------------------------------------------------------------------------*/
+
+import {
+  ActionsUnion,
+  createAction,
+  DeepReadonly,
+  FrameworkReducer,
+  FrameworkState,
+} from "@bentley/ui-framework";
+import { combineReducers, createStore, Store } from "redux";
```
[the remark this answers is not in the document]
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/frontend/app/AppState.ts:13]: I would rather use rtk if we are changing it at all. I'll take a look

File: packages/apps/desktop-viewer-test/src/frontend/components/AppComponent.tsx [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+import { SnapshotSelectFrontstage } from "./frontstages/SnapshotSelectFrontstage";
+
+type LocalStoreName = "snapshot" | "imodel" | "project";
+const makeLocalStoreName = (name: LocalStoreName) => `DesktopViewer:${name}`;
+const fromLocalStorage = (name: LocalStoreName) =>
+  window.localStorage.getItem(makeLocalStoreName(name)) ?? undefined; // replace null with undefined
+const toLocalStorage = (baseName: LocalStoreName, val?: string) => {
+  const name = makeLocalStoreName(baseName);
+  const storage = window.localStorage;
+  if (val === undefined || val.length === 0) {
```
[the remark this answers is not in the document]
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/frontend/components/AppComponent.tsx:38]: pretty sure this was initially designed to allow setting empty string, but I don't see a place where that can occur

File: packages/apps/desktop-viewer-test/src/frontend/components/AppComponent.tsx (hunk 15 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/frontend/components/AppComponent.tsx:112]: I feel like app state and ls are mutually exclusive. Redux is used because these values are set in frontstages that are "disconnected" from the component tree for all intents and purposes. Ls is used to retain that last opened model for the next time that the app is opened

File: packages/apps/desktop-viewer-test/src/frontend/components/Properties.tsx [hunk 26]
```diff
@@ -0,0 +1,50 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { IModelConnection } from "@bentley/imodeljs-frontend";
```
[the remark this answers is not in the document]
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/frontend/components/Properties.tsx]: deleted

File: packages/modules/desktop-viewer-react/src/services/Initializer.ts (hunk 9 above)
Comment (aruniverse) [reply packages/modules/desktop-viewer-react/src/services/Initializer.ts:9]: @calebmshafer heads up, we'll want to fix this core. unless there was a reason for this to begin with

**Review by aruniverse — APPROVED** [review]
lgtm! :)

File: packages/apps/desktop-viewer-test/src/backend/main.ts [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+ * that starts from the file "index.ts". That launches the viewer frontend (IModelApp).
+ */
+const viewerMain = async () => {
+  // Setup logging immediately to pick up any logging during IModelHost.startup()
+  Logger.initializeToConsole();
+  Logger.setLevelDefault(LogLevel.Warning);
+  Logger.setLevel(AppLoggerCategory.Backend, LogLevel.Info);
+
+  const electronHost: ElectronHostOptions = {
+    webResourcesPath: path.join(__dirname, "..", "..", "build"),
```
[the remark this answers is not in the document]
Comment (aruniverse) [reply packages/apps/desktop-viewer-test/src/backend/main.ts:32]: strings are showing up localized/psuedolocalized for me now. idk what changed. @calebmshafer im curious if its going to work for you.

File: packages/apps/desktop-viewer-test/package.json (hunk 12 above)
Comment (calebmshafer) [reply packages/apps/desktop-viewer-test/package.json:26]: With it being a test app, this is fine with me but as part of the CRA template I'm more unsure of if we want to continue down that path.  Let me check in on the status of the installer work for electron forge and see if we can get someone working on it.

File: packages/apps/desktop-viewer-test/src/frontend/components/frontstages/SignInFrontstage.tsx [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+  // user chose to work offline from the sign in page
+  private _onWorkOffline = async () => {
+    const frontstageDef =
+      FrontstageManager.findFrontstageDef("SnapshotSelector");
+    await FrontstageManager.setActiveFrontstageDef(frontstageDef);
+  };
+
+  private _onRegister = () => {
+    window.open(
+      "https://www.imodeljs.org/getting-started/#developer-registration",
```
Comment (calebmshafer) [inline packages/apps/desktop-viewer-test/src/frontend/components/frontstages/SignInFrontstage.tsx:49]: can you update this to point to https://developer.bentley.com/register/

File: packages/modules/desktop-viewer-react/src/services/Initializer.ts (hunk 9 above)
Comment (calebmshafer) [reply packages/modules/desktop-viewer-react/src/services/Initializer.ts:9]: Yes, the electron-manager package is switched to handle both the frontend and backend dependencies so there isn't one single barrel file, rather two "barrels" that you need to use.

File: packages/modules/desktop-viewer-react/src/services/Initializer.ts [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+      if (DesktopInitializer._reject) {
+        DesktopInitializer._reject();
+      }
+      ElectronApp.shutdown().catch(() => {
+        // Do nothing, its possible that we never started.
+      });
+    }
+  };
+
+  /** Web viewer startup */
```
Comment (calebmshafer) [inline packages/modules/desktop-viewer-react/src/services/Initializer.ts:37]: nit: comment is wrong

Comment (kckst8) [reply packages/modules/desktop-viewer-react/src/services/Initializer.ts]: nice catch. fixed

File: packages/apps/desktop-viewer-test/src/frontend/components/frontstages/SignInFrontstage.tsx (hunk 28 above)
Comment (kckst8) [reply packages/apps/desktop-viewer-test/src/frontend/components/frontstages/SignInFrontstage.tsx]: done
